# **🌟 Exercise 1 : Exploratory Data Analysis**

In [47]:
import pandas as pd
data=pd.read_csv('dataset_heart.csv')

In [48]:
data.head()

age  sex   chest pain type  resting blood pressure  serum cholestoral  \
0   70     1                4                     130                322   
1   67     0                3                     115                564   
2   57     1                2                     124                261   
3   64     1                4                     128                263   
4   74     0                2                     120                269   

   fasting blood sugar  resting electrocardiographic results  max heart rate  \
0                    0                                     2             109   
1                    0                                     2             160   
2                    0                                     0             141   
3                    0                                     0             105   
4                    0                                     2             121   

   exercise induced angina  oldpeak  ST segment  major vessels  thal  \
0                        0      2.4           2              3     3   
1                        0      1.6           2              0     7   
2                        0      0.3           1              0     7   
3                        1      0.2           2              1     7   
4                        1      0.2           1              1     3   

   heart disease  
0              2  
1              1  
2              2  
3              1  
4              1

In [49]:
from sklearn.model_selection import train_test_split

y=data['heart disease']
y=y-1

X=data.drop('heart disease',axis=1)

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)
print(X_train.describe())
print(X_test.describe())

(189, 13)
(81, 13)
(189,)
(81,)
              age        sex   chest pain type  resting blood pressure  \
count  189.000000  189.000000       189.000000              189.000000   
mean    54.751323    0.693122         3.211640              131.809524   
std      9.132393    0.462423         0.966278               17.952233   
min     29.000000    0.000000         1.000000               94.000000   
25%     48.000000    0.000000         3.000000              120.000000   
50%     56.000000    1.000000         4.000000              130.000000   
75%     62.000000    1.000000         4.000000              140.000000   
max     74.000000    1.000000         4.000000              200.000000   

       serum cholestoral  fasting blood sugar  \
count         189.000000           189.000000   
mean          251.529101             0.137566   
std            52.554890             0.345359   
min           141.000000             0.000000   
25%           215.000000             0.000000   
50%    

# **We replace heart disease to 1 and 0 instead of 2 and 1 so we can analyze and apply Logistic Regression**

# **🌟 Exercise 2 : Logistic Regression without Grid Search**

In [50]:
from sklearn.metrics import accuracy_score

from sklearn.linear_model import LogisticRegression

log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

y_pred = log_reg.predict(X_test)

print('Accuracy of logistic regression',{accuracy_score(y_test, y_pred)})


Accuracy of logistic regression {0.8641975308641975}


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


# **This model of logistic regression has accuracy of 86%, which means it predicted good that percentage of the test. But accuracy is not very important in a model like this, because it can mean that detected 86% good because 86% didnt have heart disease**

In [51]:
from sklearn.model_selection import cross_val_score

cv_scores = cross_val_score(LogisticRegression(), X, y, cv=5)
print(f"Cross-validation accuracy scores: {cv_scores}")
print(f"Mean cross-validation accuracy: {cv_scores.mean()}")

Cross-validation accuracy scores: [0.83333333 0.85185185 0.85185185 0.85185185 0.87037037]
Mean cross-validation accuracy: 0.8518518518518519


/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/usr/local/lib/python3.11/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _c

# **We divided the data in 5 and repeat the process. We can see the mean is 85% accuracy**

# **🌟 Exercise 3 : Logistic Regression with Grid Search**

In [52]:
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import GridSearchCV

scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

param_grid = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear']}


log_reg = LogisticRegression()
grid_search = GridSearchCV(log_reg, param_grid, cv=5)
grid_search.fit(X_train, y_train)

#best model
best_log_reg = grid_search.best_estimator_
y_pred = best_log_reg.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print(f"LR Accuracy (with Grid): {accuracy}")

# best hyperparameters
print(f"best hyperparameters: {grid_search.best_params_}")

LR Accuracy (with Grid): 0.8271604938271605
best hyperparameters: {'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}


# **After the model was optimized with grid search accuracy is now 82%. We are not giving this high relevance because accuracy is not the best metric to evaluate the model. We will analyze some other metrics like precision or recall later on**

# **smaller C means the model is simpler and less flexible, and big C is more  flexible and has more capacity to learn. Our C is pretty small which means that is not very flexible but also not overfitting. Is using liblinear cause is usefull for small/medium dataset. L2 gives weight to features, if weight is too high it does weight^2, so the model wont be overfitt to train data**

# **🌟 Exercise 4 : SVM without Grid Search**

In [53]:
from sklearn.svm import SVC

svm_clf = SVC(kernel='linear', C=1.0)
svm_clf.fit(X_train, y_train)

y_pred = svm_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print(f"SVM accuracy: {accuracy:.2f}")

SVM accuracy: 0.84


## **The accuracy is similar to what we obtained before, altough a bit smaller than with logistic regression. Is still not the best metric to test a model with binary target such as heart disease yes or no.**

# **🌟 Exercise 5 : SVM with Grid Search**

In [54]:
param_grid = {'C': [0.001, 0.01, 0.1, 1, 10, 100, 1000],
    'kernel': ['linear', 'rbf', 'poly'],
    'gamma': ['scale', 'auto']}

# **We are trying a few values in C to see which gives better result to not overfitt the model. In kernel we give different ways of finding a separation between classes of values. and gamma affect the area of decision (distance or margin between the line and values taken)**

In [55]:
grid_search = GridSearchCV(SVC(), param_grid, cv=5)
grid_search.fit(X_train, y_train)

# best model
best_svm_clf = grid_search.best_estimator_
y_pred = best_svm_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print(f"SVM accuracy: {accuracy:.2f}")
print(f"best parameters: {grid_search.best_params_}")

SVM accuracy: 0.83
best parameters: {'C': 1, 'gamma': 'scale', 'kernel': 'rbf'}


# **The value of C is in the middle of the values we gave so we can say there is a good balance between adjusting the model without overfitting it. Gamma scale works better for this model, and the kernel rbf because is non linear and flexible with data that is no linearly separable**

# **🌟 Exercise 6 : XGBoost without Grid Search**

In [60]:
import xgboost as xgb

xgb_clf = xgb.XGBClassifier(
    learning_rate=0.3,
    n_estimators=80,
    max_depth=6,
    subsample=0.7,
    colsample_bytree=0.7,
    random_state=100
)

xgb_clf.fit(X_train, y_train)

y_pred = xgb_clf.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)

print(f"XGBoost Accuracy: {accuracy:.2f}")


XGBoost Accuracy: 0.80


# **🌟 Exercise 7 : XGBoost with Grid Search**

In [57]:
param_grid = {
    'learning_rate': [0.01, 0.1, 0.2],
    'n_estimators': [50, 100, 200],
    'max_depth': [3, 4, 5],
    'subsample': [0.8, 0.9, 1.0],
    'colsample_bytree': [0.8, 0.9, 1.0]
}

xgb_clf = xgb.XGBClassifier()

grid_search = GridSearchCV(xgb_clf, param_grid, cv=5, n_jobs=-1, verbose=1)

grid_search.fit(X_train, y_train)

#best estimator
best_xgb_clf = grid_search.best_estimator_

y_pred = best_xgb_clf.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"accuracy: {accuracy}")

Fitting 5 folds for each of 243 candidates, totalling 1215 fits
XGBoost Accuracy with Grid Search: 0.8148148148148148


# **Here we also divided the data in 5 groups. 1 will be the test and the rest the training and will repeat the process 5 times. We are testing 243 different combinations of parameters(Hyperparameters) to check which combinations are better to predict our target. Is better to use grid cause is more precise and we dont know which combinations of parameters could predict better our target. We can see here accuracy is a bit higher than doing manually so we reinforce that using grid is more efficient**